### PR TITLE
[c10d] Ensure collectives are called with the same dtype for all tensor params.

### LINKED ISSUE #84525: `torch.distributed.all_gather` on wrong type of tensor list should raise a `TypeError`
### 🐛 Describe the bug

`torch.distributed.all_gather` should raise a `TypeError` if the tensors in `tensor_list` are not the same type as the gathered tensors. Otherwise, we end up with a silent bug where the gathered tensors are not equivalent to the tensor on each process. The following code reproduces this issue on 4 GPUs with `nccl` backend:

```python
# `gpu` refers to the current rank.

a = torch.zeros(1, 5).cuda(gpu).half() + gpu
print(f"Tensor on rank {gpu}: {a}")

gather_list = [torch.zeros(1, 5).cuda(gpu) for _ in range(dist.get_world_size())]
dist.all_gather(gather_list, a)
if gpu == 0:
    print(f"Collected tensor on rank 0:")
    for i in range(dist.get_world_size()):
        print(f"From GPU {i}: {gather_list[i]}")
```

The output is as follows: 
```
Tensor on rank 0: tensor([[0., 0., 0., 0., 0.]], device='cuda:0', dtype=torch.float16)
Tensor on rank 3: tensor([[3., 3., 3., 3., 3.]], device='cuda:3', dtype=torch.float16)
Tensor on rank 1: tensor([[1., 1., 1., 1., 1.]], device='cuda:1', dtype=torch.float16)
Tensor on rank 2: tensor([[2., 2., 2., 2., 2.]], device='cuda:2', dtype=torch.float16)

From GPU 0: tensor([[0.0000, 0.0000, 0.0078, 0.0078, 0.0078]], device='cuda:0')
From GPU 1: tensor([[ 2.0039,  2.0039, 32.0625, 32.0645, 32.0645]], device='cuda:0')
From GPU 2: tensor([[0., 0., 0., 0., 0.]], device='cuda:0')
From GPU 3: tensor([[0., 0., 0., 0., 0.]], device='cuda:0')
```

In the code above, I've initialized the `gather_list` with `fp32` (or `.float()`), while the actual gathered tensors are in `fp16` (or `.half()`).  As can be seen by the outputs, the gathered tensor is skewed, but PyTorch does not raise any errors/warnings regarding this, which makes it a silent bug. I suspect this occurs because `fp16` tensors are being read as `fp32` types, but I still think such code should raise an error, or at least a warning?

I've tested this on NVIDIA NGC 21.05 (PyTorch 1.9.0) and Conda 1.12.1. The output of `collect_env.py` below is from the conda environment.

### Versions

PyTorch version: 1.12.1
Is debug build: False
CUDA used to build PyTorch: 11.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.6 LTS (x86_64)
GCC version: (GCC) 9.4.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.27

Python version: 3.9.9 | packaged by conda-forge | (main, Dec 20 2021, 02:41:03)  [GCC 9.4.0] (64-bit runtime)
Python platform: Linux-4.15.0-163-generic-x86_64-with-glibc2.27
Is CUDA available: True
CUDA runtime version: 10.0.130
GPU models and configuration: 
GPU 0: NVIDIA GeForce RTX 2080 Ti
GPU 1: NVIDIA GeForce RTX 2080 Ti
GPU 2: NVIDIA GeForce RTX 2080 Ti
GPU 3: NVIDIA GeForce RTX 2080 Ti

Nvidia driver version: 510.85.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.3.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.3.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] mypy-extensions==0.4.3
[pip3] numpy==1.21.5
[pip3] pytorch-pfn-extras==0.5.5
[pip3] torch==1.12.1
[pip3] torchaudio==0.12.1
[pip3] torchmetrics==0.7.2
[pip3] torchvision==0.13.1
[conda] blas                      2.113                       mkl    conda-forge
[conda] blas-devel                3.9.0            13_linux64_mkl    conda-forge
[conda] cudatoolkit               11.6.0              hecad31d_10    conda-forge
[conda] libblas                   3.9.0            13_linux64_mkl    conda-forge
[conda] libcblas                  3.9.0            13_linux64_mkl    conda-forge
[conda] liblapack                 3.9.0            13_linux64_mkl    conda-forge
[conda] liblapacke                3.9.0            13_linux64_mkl    conda-forge
[conda] mkl                       2022.0.1           h8d4b97c_803    conda-forge
[conda] mkl-devel                 2022.0.1           ha770c72_804    conda-forge
[conda] mkl-include               2022.0.1           h8d4b97c_803    conda-forge
[conda] mypy_extensions           0.4.3            py39hf3d152e_4    conda-forge
[conda] numpy                     1.21.5           py39haac66dc_0    conda-forge
[conda] pytorch                   1.12.1          py3.9_cuda11.6_cudnn8.3.2_0    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] pytorch-pfn-extras        0.5.5                    pypi_0    pypi
[conda] torchaudio                0.12.1               py39_cu116    pytorch
[conda] torchmetrics              0.7.2              pyhd8ed1ab_0    conda-forge
[conda] torchvision               0.13.1               py39_cu116    pytorch

cc @pietern @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @SciPioneer @H-Huang @kwen2501

### PR DESCRIPTION
**Pull request by kumpera**
While passing tensors with different dtypes don't crash, they don't produce sensible results.

We see data tearing instead of casting.

It's not clear we want to support transparent casting so, for now, we fail when such input is presented.

Fixes #84525

Fixes #ISSUE_NUMBER

### CODE REVIEW

**Comment by numpee** [comment]
Thanks for addressing the issue. This would be a great QoL update

File: torch/distributed/distributed_c10d.py [hunk 1]
```diff
@@ -407,6 +409,21 @@ def _check_tensor_list(param, param_name):
             "to be of type List[torch.Tensor].".format(param_name)
         )
 
+def _as_iterable(obj) -> collections.Iterable:
```
Comment (kumpera) [inline torch/distributed/distributed_c10d.py:412]:
```suggestion
def _as_iterable(obj) -> collections.abc.Iterable:
```

**Review by rohan-varma — APPROVED** [review]
LGTM overall, but I guess it is technically BC breaking?

File: torch/distributed/distributed_c10d.py [hunk 2]
```diff
@@ -407,6 +409,26 @@ def _check_tensor_list(param, param_name):
...
 
+def _as_iterable(obj) -> collections.abc.Iterable:
+    return obj if isinstance(obj, list) else (obj,)
+
+def _ensure_all_tensors_same_dtype(*tensors) -> None:
+    last_dtype = None
+    for tensor in itertools.chain(*map(_as_iterable, tensors)):
+        tensor_dtype = tensor.dtype
+        # Mixing complex and its element type is allowed
+        if tensor_dtype.is_complex:
```
[the remark this answers is not in the document]
Comment (kumpera) [reply torch/distributed/distributed_c10d.py:420]: done.

**Pull request closed without merging**